python/cuml/cuml/manifold/umap_utils.pyx

```diff
@@ -110,6 +114,47 @@ cdef class GraphHolder:
     def __dealloc__(self):
         self.c_graph.reset(NULL)
 
+cdef class HostGraphHolder:
+    @staticmethod
+    cdef HostGraphHolder new_graph():
+        cdef HostGraphHolder graph = HostGraphHolder.__new__(HostGraphHolder)
+        graph.c_graph.reset(new host_COO())
+        return graph
+
+    cdef uintptr_t vals(self):
+        return <uintptr_t>self.get().vals()
+
+    cdef uintptr_t rows(self):
+        return <uintptr_t>self.get().rows()
+
+    cdef uintptr_t cols(self):
+        return <uintptr_t>self.get().cols()
+
+    cdef uint64_t get_nnz(self):
+        return self.get().get_nnz()
+
+    def get_scipy_coo(self):
+        def create_nonowning_numpy_array(ptr, dtype):
+            c_type = np.ctypeslib.as_ctypes_type(dtype)
+            c_pointer = ctypes.cast(ptr, ctypes.POINTER(c_type))
+            return np.ctypeslib.as_array(c_pointer, shape=(self.get_nnz(),))
+
+        vals = create_nonowning_numpy_array(self.vals(), np.float32)
+        rows = create_nonowning_numpy_array(self.rows(), np.int32)
+        cols = create_nonowning_numpy_array(self.cols(), np.int32)
+
+        graph = scipy.sparse.coo_matrix((vals.copy(), (rows.copy(), cols.copy())))
+        return graph
```

Comment:
This creates a full copy of the COO with memory allocated by numpy instead of within `libcuml`. This doubles the host memory (original host COO + new host COO), but the original should be freed momentarily after once the `HostGraphHolder` is dropped.

I don't love this, but I'm fine with this for now. Note that if we wanted, we could instead expose the COO from libcuml directly through the buffer protocol, avoiding the brief memory doubling.

Can you add a small comment/docstring on this method just noting that this returns a _copy_ currently (to help future readers).

Reply:
> the original should be freed momentarily after once the HostGraphHolder is dropped.

Exactly, the function that instantiate the `HostGraphHolder` should release it immediately (modulo the garbage collector).

> we could instead expose the COO from libcuml directly through the buffer protocol, avoiding the brief memory doubling

The COO matrix should have a predictable nnz before thresholding, so this would indeed be theoretically possible to pre-allocate the matrix on host and expose it in libcuml. However, we would like to use the `raft::host_coo_matrix` for the libcuml API and this utility can not be used in a non-owning way for now.